### ISSUE TITLE
I don't understand about Dimensionality in HMM

### ISSUE THREAD

**Issue by dibalikpohon** [mlpack/mlpack#3348]
Hi,
I was about to use the `HMM<DiscreteDistribution>` class with this following code:
```cpp
int main () {

        HMM<DiscreteDistribution> hmm(2, 4);

        arma::Row<size_t> labels;
        mlpack::data::Load("states.csv", labels);

        arma::mat observations;
        mlpack::data::Load("observations.csv", observations);

        hmm.Dimensionality() = observations.n_rows;

        hmm.Train({observations}, {labels}); 
}

```

But once I run it, I got this following error at the last line of the code:
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  observations must have same dimensionality as the DiscreteDistribution object
```

The "observations.csv" has 4 rows and 9 columns, and "states.csv" has only 1 row and 9 columns.

In advance of making the code, I tried to run `mlpack_hmm_train`:
```
mlpack_hmm_train  --type 'discrete' --states 2 --input_file observations.csv --labels_file states.csv --output_model_file models_discrete.xml
```

The "models_discrete.xml" yields `<dimensionality>4</dimensionality>` like this:
```xml
...
<data>
	<cereal_class_version>0</cereal_class_version>
	<dimensionality>4</dimensionality>
	<tolerance>1.0000000000000001e-05</tolerance>
	<transition>
		<n_rows>2</n_rows>
		<n_cols>2</n_cols>
		<vec_state>0</vec_state>
...
```

which I believe it's supposed to be the number of rows in the "observations.csv" file, so I set it.

What am I getting wrong here?

**Comment by OmkarGowda990** [comment]
We have to pass observations and probabilities in the function. While going through the code I found the error is thrown when it checks the condition if (observations.n_rows != probabilities.size())
(Make sure the observations have the same dimension as the probabilities.)

**Comment by AdarshSantoria** [comment]
Dimensionality is the the total number of features in observable states.
When you wrote "HMM<DiscreteDistribution> hmm(2, 4);", you assume 2 hidden states and 4 observable states and as default dimensionality (number of features in observable states) is 1 as not mentioned. According to your error, you have not fixed dimensionality (number of columns or features) equal to 1. Once the model is build you can't change dimensionality directly you also have to change it's other dependencies like emission matrix.

Here is an example for dimensionality = 4,
`#include <mlpack/core.hpp>
#include <mlpack/methods/hmm.hpp>
#include <bits/stdc++.h>

using namespace std;
using namespace arma;
using namespace mlpack;

int main(){
  HMM<DiscreteDistribution> hmm(2,vector<arma::vec>{"0.3 0.1 0.1 0.1 0.1",
                                                                                          "0.0 0.5 0.3 0.2 0.0",
                                                                                          "0.3 0.4 0.7 0.5 0.2",
                                                                                          "0.0 0.0 0.4 0.5 0.0"});
  // 5 observable states of dimensionality 4
  cout<<hmm.Dimensionality();
  Row<size_t> labels;
  labels.randn(100);
  cout<<labels;
  mat observations = randu<mat>(4, 100);
  hmm.Train({observations});
}`
I hope this will clear your doubt about dimensionality in hmm.

**Comment by AdarshSantoria** [comment]
Dimensionality is the the total number of features in observable states.
When you wrote "HMM hmm(2, 4);", you assume 2 hidden states and 4 observable states and as default dimensionality (number of features in observable states) is 1 as not mentioned. According to your error, you have not fixed dimensionality (number of columns or features) equal to 1. Once the model is build you can't change dimensionality directly you also have to change it's other dependencies like emission matrix.

Here is an example for dimensionality = 4,

```
#include <mlpack/core.hpp>
#include <mlpack/methods/hmm.hpp>
#include <bits/stdc++.h>

using namespace std;
using namespace arma;
using namespace mlpack;

int main(){
HMM hmm(2,vectorarma::vec{"0.3 0.1 0.1 0.1 0.1",
"0.0 0.5 0.3 0.2 0.0",
"0.3 0.4 0.7 0.5 0.2",
"0.0 0.0 0.4 0.5 0.0"});
// 5 observable states of dimensionality 4
cout<<hmm.Dimensionality();
Row<size_t> labels;
labels.randn(100);
cout<<labels;
mat observations = randu(4, 100);
hmm.Train({observations});
}
```

I hope this will clear your doubt about dimensionality in hmm.

**Comment by dibalikpohon** [comment]
Hi @AdarshSantoria, Thank you for the reply. 
If I could ask more questions, what does the feature represents in the observable states? Does each observables(emisisons) can have values?
Thank you.

**Comment by dibalikpohon** [comment]
Hi @rcurtin, thank you for the clarification on using the Dimensionality. I have set my constructor as you suggests and the code runs fine. 

I need to clarify things to get my understanding better. @AdarshSantoria says 
> Dimensionality is the the total number of features in observable states.

and you explained more about it. It gives me pretty broad knowledge about the HMM — thinking that HMM only has these three properties: initial state, hidden states, and observable states before I know about dimensionality. 

So my next question is: Does this means by setting the dimensionality  — e.g. 4  means we have  4 different things to consider and each consideration has some number of observations. Am I right?

Also, can each observable states have values, e.g. ranging from 0 to 1 in a real number?

Apologize for my lack of experience.

**Comment by rcurtin** [comment]
You're right that the HMM has three properties: initial state, hidden states, and observable states (those are called emissions by our code).  Now, the initial state is a vector of probabilities, and the hidden states are controlled by a probability transition matrix.  At each time point, the HMM, given its internal state, can emit some observed state.  But, note that the observed state is modeled as a probability distribution over the hidden state.  However, these observed states can have a number of types.

In the most simple case, the observation is a discrete integer representing some observed state.  So, e.g., if your HMM was about the weather, maybe your observed states can be `0` representing "cloudy", and `1` representing "clear", and the hidden state is whether or not it is raining.  (Maybe that is not the best example in the world, but bear with me...)  Since the observed state is a discrete state, the distribution of observed states for a given hidden state is a discrete distribution.  So, if the hidden state is rainy, maybe you will see that `P(0)` is high and `P(1)` is low; and if the hidden state is not rainy, maybe `P(0)` is low and `P(1)` is high.  Here, you would use `HMM<DiscreteDistribution>`.

In that weather example, the discrete observed states are one-dimensional (that is, they are a scalar: `0` or `1`).

But what if the observed state is not a discrete quantity but instead some real number?  For instance, to work on the weather example further, maybe the observation is the percentage of cloud cover (some number between 0 and 1).  Now, the discrete distribution is no longer reasonable to use, since the observed state is a continuous variable.  Instead, we have to use some continuous distribution.  Generally, this might be `HMM<GaussianDistribution>`, but you could also use Gaussian mixture models: `HMM<GMM>`, or really any other probability distribution that mlpack implements.

I still haven't gotten to your actual question about dimensionality but I finally will now: in the examples above, the observations were all one-dimensional (they were scalars).  But you could imagine that observations can be multi-dimensional.  Take the weather example again---what if what we observe is not just percentage of cloud cover, but a 3-dimensional vector whose first component is percentage of cloud cover, second component is temperature, and third component is wind speed?  Now the `GaussianDistribution` that we fit to each hidden state is three-dimensional, not one-dimensional.

There is an analog for discrete observations too; you can have multi-dimensional discrete distributions.  Using the weather example, maybe we observe a 3-dimensional discrete vector, where the first element is whether or not there are clouds, the second element is `0` if the temperature is "cold" and `1` if it is "hot" (for some definition thereof), and the third is `0` if there is no wind, and `1` otherwise.  (These I guess are just discretized versions of the continuous variables in the paragraph above.)  In this case, the dimensionality of each `DiscreteDistribution` would be 3, if you were to model this problem with `HMM<DiscreteDistribution>`.

In conclusion, the dimensionality of the observations is something that's specific to the emissions/observable states, and doesn't actually have to do with the internals of the HMM (the initial state or hidden state transition matrix).  I hope the examples I gave are illuminating---but if not, I can try again, just let me know. :)

**Comment by dibalikpohon** [comment]
Ooh, I now understand! Thank you for taking your time to explain. I greatly appreciate your help.

Karunia.
